packages/nimble-components/src/table-column/mixins/custom-sort-order.ts

```diff
@@ -0,0 +1,96 @@
+type CustomSortOrderTableColumn<
+    TColumnValidator extends ColumnValidator<['invalidCustomSortWithGrouping']>
+> = Pick<TableColumn<unknown, TColumnValidator>, 'columnInternals'>;
+// prettier-ignore
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+type CustomSortOrderTableColumnConstructor<TColumnValidator extends ColumnValidator<['invalidCustomSortWithGrouping']>> = abstract new (...args: any[]) => CustomSortOrderTableColumn<TColumnValidator>;
+
+// As the returned class is internal to the function, we can't write a signature that uses is directly, so rely on inference
+// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
+export function mixinCustomSortOrderColumnAPI<
+    TBase extends CustomSortOrderTableColumnConstructor<TColumnValidator>,
+    TColumnValidator extends ColumnValidator<['invalidCustomSortWithGrouping']>
+>(base: TBase) {
+    /**
+     * The Mixin that provides a concrete column with the API to allow sorting
+     * by a field other than the default for that column.
+     */
+    abstract class CustomSortOrderColumn extends base {
+        public sortByFieldName?: string;
+
+        /** @internal */
+        public notifier?: Notifier;
+
+        /** @internal */
+        public sortByFieldNameChanged(): void {
+            this.updateOperandDataRecordFieldName();
+
+            if (typeof this.sortByFieldName === 'string' && !this.notifier) {
+                this.notifier = Observable.getNotifier(this.columnInternals);
+                this.notifier.subscribe(this);
+                this.updateCustomColumnSortingValidity();
+            } else {
+                this.notifier?.unsubscribe(this);
+                this.notifier = undefined;
+                this.updateCustomColumnSortingValidity();
+            }
+        }
+
+        /** @internal */
+        public updateOperandDataRecordFieldName(): void {
+            if (typeof this.sortByFieldName === 'string') {
+                this.columnInternals.operandDataRecordFieldName = this.sortByFieldName;
+                this.columnInternals.sortOperation = TableColumnSortOperation.basic;
+            } else {
+                this.columnInternals.operandDataRecordFieldName = this.getDefaultSortFieldName();
+                this.columnInternals.sortOperation = this.getDefaultSortOperation();
+            }
+        }
+
+        /** @internal */
+        public handleChange(_source: unknown, args: unknown): void {
+            if (args === 'groupingDisabled' || args === 'groupIndex') {
+                this.updateCustomColumnSortingValidity();
+            } else if (args === 'dataRecordFieldNames') {
+                this.updateOperandDataRecordFieldName();
+            }
+        }
+
+        /** @internal */
+        public updateCustomColumnSortingValidity(): void {
```

Comment:
Corner case to consider: a custom sort order but sorting is disabled. Should grouping be allowed in this case? It's admittedly a weird configuration, but a couple (weak) arguments in favor of supporting it:
1. we check whether grouping is disabled as part of the validity calculation so checking sorting feels symmetric
2. I can imagine a client wanting to toggle between custom sorting and grouping while leaving the configuration for both of those states set the whole time

If you see arguments against considering this configuration as valid, I don't feel strongly about supporting it. We could always add support if a client requests it.

Reply:
Chatted with molly offline and the issue is that the operand config is used in both sorting and grouping so even though sorting may be temporary disabled, the issue is the same in that the right way to apply a configured sortByFieldName to grouping.

Reply:
I've reverted the change I initially made in response to this discussion thread. If `sortByFieldName` is configured with grouping, the column is invalid, regardless of whether or not sorting is enabled or disabled.